## 5.索引、选取和过滤

In [1]:
import numpy as np
import pandas as pd
from pandas import Series
from pandas import DataFrame

In [2]:
obj = Series(np.arange(4.),index = ['a','b','c','d'])
obj

a    0.0
b    1.0
c    2.0
d    3.0
dtype: float64

In [3]:
obj['a']

0.0

In [4]:
obj[0]

0.0

In [5]:
obj[1:2]

b    1.0
dtype: float64

In [6]:
obj[['b','a']]

b    1.0
a    0.0
dtype: float64

In [7]:
obj[[0,1]]

a    0.0
b    1.0
dtype: float64

In [8]:
obj[obj<2]

a    0.0
b    1.0
dtype: float64

In [9]:
obj['b':'c']

b    1.0
c    2.0
dtype: float64

In [10]:
obj['b':'c'] = 5
obj

a    0.0
b    5.0
c    5.0
d    3.0
dtype: float64

In [11]:
data = DataFrame(np.arange(16).reshape((4,4)),index = ['Ohio','Colorado','Utah','New York'],
                 columns = ['one','two','three','four'])
data

one  two  three  four
Ohio        0    1      2     3
Colorado    4    5      6     7
Utah        8    9     10    11
New York   12   13     14    15

In [12]:
data['two']

Ohio         1
Colorado     5
Utah         9
New York    13
Name: two, dtype: int32

In [13]:
data[['two','three']]

two  three
Ohio        1      2
Colorado    5      6
Utah        9     10
New York   13     14

In [14]:
data[:1]

one  two  three  four
Ohio    0    1      2     3

In [15]:
data[1:]

one  two  three  four
Colorado    4    5      6     7
Utah        8    9     10    11
New York   12   13     14    15

In [16]:
data[data['three']>5]

one  two  three  four
Colorado    4    5      6     7
Utah        8    9     10    11
New York   12   13     14    15

In [17]:
data<5

one    two  three   four
Ohio       True   True   True   True
Colorado   True  False  False  False
Utah      False  False  False  False
New York  False  False  False  False

In [18]:
data[data<5]

one  two  three  four
Ohio      0.0  1.0    2.0   3.0
Colorado  4.0  NaN    NaN   NaN
Utah      NaN  NaN    NaN   NaN
New York  NaN  NaN    NaN   NaN

In [19]:
data[data<5] = 0
data

one  two  three  four
Ohio        0    0      0     0
Colorado    0    5      6     7
Utah        8    9     10    11
New York   12   13     14    15

In [20]:
data.ix['Colorado',['two','three']]

c:\users\administrator\appdata\local\programs\python\python37\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  """Entry point for launching an IPython kernel.


two      5
three    6
Name: Colorado, dtype: int32

In [21]:
data.ix[['Colorado','Utah'],[3,0,1]]

c:\users\administrator\appdata\local\programs\python\python37\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  """Entry point for launching an IPython kernel.


four  one  two
Colorado     7    0    5
Utah        11    8    9

In [22]:
data.ix[2]

c:\users\administrator\appdata\local\programs\python\python37\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  """Entry point for launching an IPython kernel.


one       8
two       9
three    10
four     11
Name: Utah, dtype: int32

In [23]:
data.ix[:'Utah',1]

c:\users\administrator\appdata\local\programs\python\python37\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  """Entry point for launching an IPython kernel.


Ohio        0
Colorado    5
Utah        9
Name: two, dtype: int32

In [24]:
data.ix[:'Utah','two']

c:\users\administrator\appdata\local\programs\python\python37\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  """Entry point for launching an IPython kernel.


Ohio        0
Colorado    5
Utah        9
Name: two, dtype: int32

In [25]:
data.ix[data.three>5,:3]

c:\users\administrator\appdata\local\programs\python\python37\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  """Entry point for launching an IPython kernel.


one  two  three
Colorado    0    5      6
Utah        8    9     10
New York   12   13     14

![jupyter](./8.10.1.png)

## 6.算术运算和数据对齐

In [26]:
s1 = Series([7.3,-2.5,3.4,1.5],index = ['a','c','d','e'])
s2 = Series([-2.1,3.6,-1.5,4,3.1],index = ['a','c','e','f','g'])

In [27]:
s1

a    7.3
c   -2.5
d    3.4
e    1.5
dtype: float64

In [28]:
s2

a   -2.1
c    3.6
e   -1.5
f    4.0
g    3.1
dtype: float64

In [29]:
s1+s2

a    5.2
c    1.1
d    NaN
e    0.0
f    NaN
g    NaN
dtype: float64

In [30]:
df1 = DataFrame(np.arange(9.).reshape((3,3)),columns = list('bcd'),index = ['Ohio','Texas','Colorado'])
df2 = DataFrame(np.arange(12.).reshape((4,3)),columns = list('bde'),index = ['Utah','Ohio','Texas','Oregon'])

In [31]:
df1

b    c    d
Ohio      0.0  1.0  2.0
Texas     3.0  4.0  5.0
Colorado  6.0  7.0  8.0

In [32]:
df2

b     d     e
Utah    0.0   1.0   2.0
Ohio    3.0   4.0   5.0
Texas   6.0   7.0   8.0
Oregon  9.0  10.0  11.0

In [33]:
df1+df2

b   c     d   e
Colorado  NaN NaN   NaN NaN
Ohio      3.0 NaN   6.0 NaN
Oregon    NaN NaN   NaN NaN
Texas     9.0 NaN  12.0 NaN
Utah      NaN NaN   NaN NaN

## 7.在算术方法中填充数值

In [34]:
df1 = DataFrame(np.arange(12.).reshape((3,4)),columns = list('abcd'))
df2 = DataFrame(np.arange(20.).reshape((4,5)),columns = list('abcde'))

In [35]:
df1

a    b     c     d
0  0.0  1.0   2.0   3.0
1  4.0  5.0   6.0   7.0
2  8.0  9.0  10.0  11.0

In [36]:
df2

a     b     c     d     e
0   0.0   1.0   2.0   3.0   4.0
1   5.0   6.0   7.0   8.0   9.0
2  10.0  11.0  12.0  13.0  14.0
3  15.0  16.0  17.0  18.0  19.0

In [37]:
df1+df2

a     b     c     d   e
0   0.0   2.0   4.0   6.0 NaN
1   9.0  11.0  13.0  15.0 NaN
2  18.0  20.0  22.0  24.0 NaN
3   NaN   NaN   NaN   NaN NaN

In [38]:
df1.add(df2,fill_value = 0)

a     b     c     d     e
0   0.0   2.0   4.0   6.0   4.0
1   9.0  11.0  13.0  15.0   9.0
2  18.0  20.0  22.0  24.0  14.0
3  15.0  16.0  17.0  18.0  19.0

In [39]:
df1.reindex(columns = df2.columns)

a    b     c     d   e
0  0.0  1.0   2.0   3.0 NaN
1  4.0  5.0   6.0   7.0 NaN
2  8.0  9.0  10.0  11.0 NaN

In [40]:
df1.reindex(columns = df2.columns,fill_value = 0)

a    b     c     d  e
0  0.0  1.0   2.0   3.0  0
1  4.0  5.0   6.0   7.0  0
2  8.0  9.0  10.0  11.0  0

![jupyter](./8.10.2.png)

## 8. DataFrame和Series之间的运算

In [41]:
arr = np.arange(12.).reshape((3,4))
arr

array([[ 0.,  1.,  2.,  3.],
       [ 4.,  5.,  6.,  7.],
       [ 8.,  9., 10., 11.]])

In [42]:
arr[0]

array([0., 1., 2., 3.])

In [43]:
arr-arr[0]

array([[0., 0., 0., 0.],
       [4., 4., 4., 4.],
       [8., 8., 8., 8.]])

In [44]:
frame = DataFrame(np.arange(12.).reshape((4,3)),columns = list('bde'),index = ['Utah','Ohio','Texas','Oregon'])

In [45]:
frame

b     d     e
Utah    0.0   1.0   2.0
Ohio    3.0   4.0   5.0
Texas   6.0   7.0   8.0
Oregon  9.0  10.0  11.0

In [46]:
series = frame.ix[0]
series

c:\users\administrator\appdata\local\programs\python\python37\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  """Entry point for launching an IPython kernel.


b    0.0
d    1.0
e    2.0
Name: Utah, dtype: float64

In [47]:
frame - series

b    d    e
Utah    0.0  0.0  0.0
Ohio    3.0  3.0  3.0
Texas   6.0  6.0  6.0
Oregon  9.0  9.0  9.0

In [48]:
series2 = Series(range(3),index = ['b','e','f'])
series2

b    0
e    1
f    2
dtype: int64

In [49]:
frame + series2

b   d     e   f
Utah    0.0 NaN   3.0 NaN
Ohio    3.0 NaN   6.0 NaN
Texas   6.0 NaN   9.0 NaN
Oregon  9.0 NaN  12.0 NaN

 匹配行并且在列上进行广播

In [50]:
frame

b     d     e
Utah    0.0   1.0   2.0
Ohio    3.0   4.0   5.0
Texas   6.0   7.0   8.0
Oregon  9.0  10.0  11.0

In [51]:
series3 = frame['d']
series3

Utah       1.0
Ohio       4.0
Texas      7.0
Oregon    10.0
Name: d, dtype: float64

In [52]:
frame.sub(series3,axis = 0)

b    d    e
Utah   -1.0  0.0  1.0
Ohio   -1.0  0.0  1.0
Texas  -1.0  0.0  1.0
Oregon -1.0  0.0  1.0

## 9.函数应用和映射

numpy的ufuns（元素级数组方法）也可以应用于操作pandas对象

In [53]:
frame = DataFrame(np.random.randn(4,3),columns = list('bde'),index = ['Utah','Ohio','Texas','Oregon'])

In [54]:
frame

b         d         e
Utah   -0.666673  1.154003  1.295973
Ohio   -1.022062 -1.284811 -1.371522
Texas   1.755952 -1.386168  1.146424
Oregon  0.412676 -0.864467 -0.226821

In [55]:
np.abs(frame)

b         d         e
Utah    0.666673  1.154003  1.295973
Ohio    1.022062  1.284811  1.371522
Texas   1.755952  1.386168  1.146424
Oregon  0.412676  0.864467  0.226821

In [56]:
f = lambda x:x.max() - x.min()

In [57]:
frame.apply(f)

b    2.778014
d    2.540171
e    2.667495
dtype: float64

In [58]:
frame.apply(f,axis = 1)

Utah      1.962646
Ohio      0.349460
Texas     3.142120
Oregon    1.277143
dtype: float64

In [59]:
def f(x):
    return Series([x.min(),x.max()],index = ['min','max'])

In [60]:
frame.apply(f)

b         d         e
min -1.022062 -1.386168 -1.371522
max  1.755952  1.154003  1.295973

## 10.排序和排名

In [61]:
obj = Series(range(4),index = ['d','a','b','c'])
obj

d    0
a    1
b    2
c    3
dtype: int64

In [62]:
obj.sort_index()

a    1
b    2
c    3
d    0
dtype: int64

In [63]:
obj.sort_values()

d    0
a    1
b    2
c    3
dtype: int64

In [64]:
obj = Series([4,np.nan,7,8,np.nan])
obj

0    4.0
1    NaN
2    7.0
3    8.0
4    NaN
dtype: float64

In [65]:
obj.sort_values()

0    4.0
2    7.0
3    8.0
1    NaN
4    NaN
dtype: float64

In [66]:
frame = DataFrame(np.arange(8).reshape((2,4)),index = ['three','one'],columns = ['d','a','b','c'])
frame

d  a  b  c
three  0  1  2  3
one    4  5  6  7

In [67]:
frame.sort_index()

d  a  b  c
one    4  5  6  7
three  0  1  2  3

In [68]:
frame.sort_index(axis = 1)

a  b  c  d
three  1  2  3  0
one    5  6  7  4

In [69]:
frame.sort_index(ascending = False)

d  a  b  c
three  0  1  2  3
one    4  5  6  7

In [70]:
frame.sort_index(axis =1,ascending = False)

d  c  b  a
three  0  3  2  1
one    4  7  6  5

In [71]:
frame = DataFrame({'b':[4,7,-3,2],'a':[0,1,0,1]})
frame

b  a
0  4  0
1  7  1
2 -3  0
3  2  1

In [72]:
frame.sort_values(by = 'b')

b  a
2 -3  0
3  2  1
0  4  0
1  7  1

In [73]:
obj = Series([7,-5,7,4,2,0,4])
obj

0    7
1   -5
2    7
3    4
4    2
5    0
6    4
dtype: int64

rank() 为各组分配一个平均排名

In [74]:
obj.rank()

0    6.5
1    1.0
2    6.5
3    4.5
4    3.0
5    2.0
6    4.5
dtype: float64

rank(method = 'first')表示按照顺序给出排名 ，没有计算平均

In [75]:
obj.rank(method = 'first')

0    6.0
1    1.0
2    7.0
3    4.0
4    3.0
5    2.0
6    5.0
dtype: float64

In [76]:
obj.rank(ascending = False,method = 'max')

0    2.0
1    7.0
2    2.0
3    4.0
4    5.0
5    6.0
6    4.0
dtype: float64

In [77]:
frame = DataFrame({'b':[4.3,7,-3,2],'a':[0,1,0,1],'c':[-2,5,8,-2.5]})
frame

b  a    c
0  4.3  0 -2.0
1  7.0  1  5.0
2 -3.0  0  8.0
3  2.0  1 -2.5

In [78]:
frame.rank(axis = 1)

b    a    c
0  3.0  2.0  1.0
1  3.0  1.0  2.0
2  1.0  2.0  3.0
3  3.0  2.0  1.0

![jupyter](./8.11.1.png)

## 11.带有重复值的轴索引

In [79]:
obj = Series(range(5),index = ['a','a','b','b','c'])
obj

a    0
a    1
b    2
b    3
c    4
dtype: int64

索引is_unique的属性可以表示它的值是否唯一

In [80]:
obj.index.is_unique

False

In [81]:
obj['a']

a    0
a    1
dtype: int64

In [82]:
obj['c']

4

In [83]:
df = DataFrame(np.random.randn(4,3),index = ['a','a','b','b'])
df

0         1         2
a -1.182444 -0.081084  0.366714
a  0.963483  1.106887 -0.312127
b  0.926207 -1.448255 -1.102677
b -0.606663 -0.067706  0.384065

In [84]:
df.ix['a']

c:\users\administrator\appdata\local\programs\python\python37\lib\site-packages\ipykernel_launcher.py:1: DeprecationWarning: 
.ix is deprecated. Please use
.loc for label based indexing or
.iloc for positional indexing

See the documentation here:
http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated
  """Entry point for launching an IPython kernel.


0         1         2
a -1.182444 -0.081084  0.366714
a  0.963483  1.106887 -0.312127

## 12. 汇总和计算描述统计

In [85]:
df = DataFrame([[1.4,np.nan],[7.1,-4.5],[np.nan,np.nan],[0.75,-1.3]],index = ['a','b','c','d'],columns = ['one','two'])
df

one  two
a  1.40  NaN
b  7.10 -4.5
c   NaN  NaN
d  0.75 -1.3

In [86]:
df.sum()

one    9.25
two   -5.80
dtype: float64

In [87]:
df.sum(axis = 1)

a    1.40
b    2.60
c    0.00
d   -0.55
dtype: float64

In [88]:
df.mean(axis = 1,skipna = False)

a      NaN
b    1.300
c      NaN
d   -0.275
dtype: float64

![jupyter](./8.11.2.png)

In [89]:
df.idxmax()

one    b
two    d
dtype: object

cumsum()累加

In [90]:
df.cumsum()

one  two
a  1.40  NaN
b  8.50 -4.5
c   NaN  NaN
d  9.25 -5.8

In [91]:
df.describe()

one       two
count  3.000000  2.000000
mean   3.083333 -2.900000
std    3.493685  2.262742
min    0.750000 -4.500000
25%    1.075000 -3.700000
50%    1.400000 -2.900000
75%    4.250000 -2.100000
max    7.100000 -1.300000

1.count:非空值总数

2.mean:非空值的平均值

3.std:方差

4.min:最小值

5.25%：25%分位数

6.50%：50%分位数（中位数）

7.75%:  75%分位数

8.max: 最大值

## 13.相关系数与协方差

In [92]:
import pandas_datareader.data as web

## 14.唯一值、值计数以及成员资格

In [93]:
obj = Series(['c','a','d','a','b','b','c','c'])
obj

0    c
1    a
2    d
3    a
4    b
5    b
6    c
7    c
dtype: object

In [94]:
uniques = obj.unique()
uniques

array(['c', 'a', 'd', 'b'], dtype=object)

In [95]:
obj.value_counts()

c    3
a    2
b    2
d    1
dtype: int64

In [96]:
mask =obj.isin(['b','c'])
mask

0     True
1    False
2    False
3    False
4     True
5     True
6     True
7     True
dtype: bool

In [97]:
obj[mask]

0    c
4    b
5    b
6    c
7    c
dtype: object

![jupyter](./8.11.3.png)

In [98]:
data = DataFrame({'qu1':[1,3,4,3,4],
                 'qu2':[2,3,1,2,3],
                 'qu3':[1,5,2,4,4]})
data

qu1  qu2  qu3
0    1    2    1
1    3    3    5
2    4    1    2
3    3    2    4
4    4    3    4

In [99]:
data.apply(pd.value_counts)

qu1  qu2  qu3
1  1.0  1.0  1.0
2  NaN  2.0  1.0
3  2.0  2.0  NaN
4  2.0  NaN  2.0
5  NaN  NaN  1.0

In [100]:
data.apply(pd.value_counts).fillna(0)

qu1  qu2  qu3
1  1.0  1.0  1.0
2  0.0  2.0  1.0
3  2.0  2.0  0.0
4  2.0  0.0  2.0
5  0.0  0.0  1.0

## 15.处理缺失数据

In [101]:
string_data = Series(['aardvark','artichoke',np.nan,'avocado'])
string_data

0     aardvark
1    artichoke
2          NaN
3      avocado
dtype: object

In [102]:
string_data.isnull()

0    False
1    False
2     True
3    False
dtype: bool

In [103]:
string_data[0] = None
string_data.isnull()

0     True
1    False
2     True
3    False
dtype: bool

![jupyter](./8.11.4.png)

## 16.滤除缺失数据

In [104]:
from numpy import nan as NA

In [105]:
data = Series([1,NA,3.5,NA,7])
data

0    1.0
1    NaN
2    3.5
3    NaN
4    7.0
dtype: float64

In [106]:
data.dropna()

0    1.0
2    3.5
4    7.0
dtype: float64

dropna返回一个仅含非空数据和索引的series

In [107]:
data[data.notnull()]

0    1.0
2    3.5
4    7.0
dtype: float64

In [108]:
data[data.isnull()]

1   NaN
3   NaN
dtype: float64

In [109]:
data = DataFrame([[1.,6.5,3.],[1.,NA,NA],[NA,NA,NA],[NA,6.5,3.]])
data

0    1    2
0  1.0  6.5  3.0
1  1.0  NaN  NaN
2  NaN  NaN  NaN
3  NaN  6.5  3.0

In [110]:
cleaned = data.dropna()
cleaned

0    1    2
0  1.0  6.5  3.0

In [111]:
data.dropna(how = 'all')

0    1    2
0  1.0  6.5  3.0
1  1.0  NaN  NaN
3  NaN  6.5  3.0

In [112]:
data[4] = NA
data

0    1    2   4
0  1.0  6.5  3.0 NaN
1  1.0  NaN  NaN NaN
2  NaN  NaN  NaN NaN
3  NaN  6.5  3.0 NaN

In [113]:
data.dropna(how = 'all',axis = 1)

0    1    2
0  1.0  6.5  3.0
1  1.0  NaN  NaN
2  NaN  NaN  NaN
3  NaN  6.5  3.0

In [114]:
df = DataFrame(np.random.randn(7,3))
df

0         1         2
0 -0.927755 -1.255639 -2.015748
1 -0.801296 -1.266911  1.603177
2 -0.597145  1.595318  1.281146
3 -0.693067 -1.484757  0.049125
4 -0.792677  2.347452 -0.345777
5  0.566035 -0.395723  0.176068
6  0.708794  0.091606  0.311640

In [115]:
df.loc[:4,1] = NA 
df.loc[:2,2] = NA
df

0         1         2
0 -0.927755       NaN       NaN
1 -0.801296       NaN       NaN
2 -0.597145       NaN       NaN
3 -0.693067       NaN  0.049125
4 -0.792677       NaN -0.345777
5  0.566035 -0.395723  0.176068
6  0.708794  0.091606  0.311640

In [116]:
df.dropna(thresh = 3)

0         1         2
5  0.566035 -0.395723  0.176068
6  0.708794  0.091606  0.311640

  thresh，该参数的类型为整数，它的作用是，比如 thresh=3，会在一行中至少有 3 个非 NA 值时将其保留。

## 17.填充缺失数据

In [117]:
df.fillna(0)

0         1         2
0 -0.927755  0.000000  0.000000
1 -0.801296  0.000000  0.000000
2 -0.597145  0.000000  0.000000
3 -0.693067  0.000000  0.049125
4 -0.792677  0.000000 -0.345777
5  0.566035 -0.395723  0.176068
6  0.708794  0.091606  0.311640

In [118]:
df.fillna({1:0.5,2:-1})

0         1         2
0 -0.927755  0.500000 -1.000000
1 -0.801296  0.500000 -1.000000
2 -0.597145  0.500000 -1.000000
3 -0.693067  0.500000  0.049125
4 -0.792677  0.500000 -0.345777
5  0.566035 -0.395723  0.176068
6  0.708794  0.091606  0.311640

In [119]:
_ = df.fillna(0,inplace = True)

In [120]:
df

0         1         2
0 -0.927755  0.000000  0.000000
1 -0.801296  0.000000  0.000000
2 -0.597145  0.000000  0.000000
3 -0.693067  0.000000  0.049125
4 -0.792677  0.000000 -0.345777
5  0.566035 -0.395723  0.176068
6  0.708794  0.091606  0.311640

In [121]:
df = DataFrame(np.random.randn(6,3))

In [122]:
df

0         1         2
0 -0.581952 -0.231465  0.256524
1  0.920218 -1.045661 -0.572126
2  0.490600  1.188068 -0.452978
3 -0.186083  0.595613 -0.975007
4 -0.858656 -0.315382 -0.243314
5 -0.627965  2.628690  1.682667

In [123]:
df.loc[2:,1] = NA ; df.loc[4:,2] = NA

In [124]:
df

0         1         2
0 -0.581952 -0.231465  0.256524
1  0.920218 -1.045661 -0.572126
2  0.490600       NaN -0.452978
3 -0.186083       NaN -0.975007
4 -0.858656       NaN       NaN
5 -0.627965       NaN       NaN

In [125]:
df.fillna(method = 'ffill')

0         1         2
0 -0.581952 -0.231465  0.256524
1  0.920218 -1.045661 -0.572126
2  0.490600 -1.045661 -0.452978
3 -0.186083 -1.045661 -0.975007
4 -0.858656 -1.045661 -0.975007
5 -0.627965 -1.045661 -0.975007

In [126]:
df.fillna(method = 'ffill',limit = 2)

0         1         2
0 -0.581952 -0.231465  0.256524
1  0.920218 -1.045661 -0.572126
2  0.490600 -1.045661 -0.452978
3 -0.186083 -1.045661 -0.975007
4 -0.858656       NaN -0.975007
5 -0.627965       NaN -0.975007

In [127]:
data = Series([1.,NA,3.5,NA,7])

In [128]:
data

0    1.0
1    NaN
2    3.5
3    NaN
4    7.0
dtype: float64

In [129]:
data.fillna(data.mean())

0    1.000000
1    3.833333
2    3.500000
3    3.833333
4    7.000000
dtype: float64

![jupyter](./8.11.5.png)

## 18. 层次化索引

In [130]:
data = Series(np.random.randn(10),index = [['a','a','a','b','b','b','c','c','d','d'],[1,2,3,1,2,3,1,2,2,3]])

In [131]:
data

a  1    1.347466
   2    0.109289
   3   -0.951960
b  1    0.138964
   2   -0.053714
   3   -1.843136
c  1    0.701553
   2   -1.374723
d  2    0.110358
   3    0.069612
dtype: float64

In [132]:
data.index

MultiIndex(levels=[['a', 'b', 'c', 'd'], [1, 2, 3]],
           codes=[[0, 0, 0, 1, 1, 1, 2, 2, 3, 3], [0, 1, 2, 0, 1, 2, 0, 1, 1, 2]])

In [133]:
data['b']

1    0.138964
2   -0.053714
3   -1.843136
dtype: float64

In [134]:
data['b':'c']

b  1    0.138964
   2   -0.053714
   3   -1.843136
c  1    0.701553
   2   -1.374723
dtype: float64

In [135]:
data.loc[['b','d']]

b  1    0.138964
   2   -0.053714
   3   -1.843136
d  2    0.110358
   3    0.069612
dtype: float64

In [136]:
data[:,2]

a    0.109289
b   -0.053714
c   -1.374723
d    0.110358
dtype: float64

In [137]:
data.unstack()

1         2         3
a  1.347466  0.109289 -0.951960
b  0.138964 -0.053714 -1.843136
c  0.701553 -1.374723       NaN
d       NaN  0.110358  0.069612

In [138]:
data.unstack().stack()

a  1    1.347466
   2    0.109289
   3   -0.951960
b  1    0.138964
   2   -0.053714
   3   -1.843136
c  1    0.701553
   2   -1.374723
d  2    0.110358
   3    0.069612
dtype: float64

In [139]:
frame = DataFrame(np.arange(12).reshape((4,3)),index = [['a','a','b','b'],[1,2,1,2]],columns = [['Ohio','Ohio','Colorado'],
                                                                                                ['Green','Red','Green']])

In [140]:
frame

Ohio     Colorado
    Green Red    Green
a 1     0   1        2
  2     3   4        5
b 1     6   7        8
  2     9  10       11

In [141]:
frame.index.names = ['key1','key2']

In [142]:
frame.columns.names = ['state','color']

In [143]:
frame

state      Ohio     Colorado
color     Green Red    Green
key1 key2                   
a    1        0   1        2
     2        3   4        5
b    1        6   7        8
     2        9  10       11

In [144]:
frame['Ohio']

color      Green  Red
key1 key2            
a    1         0    1
     2         3    4
b    1         6    7
     2         9   10

## 19.重新分级排序

In [145]:
frame.swaplevel('key1','key2')

state      Ohio     Colorado
color     Green Red    Green
key2 key1                   
1    a        0   1        2
2    a        3   4        5
1    b        6   7        8
2    b        9  10       11

## 20.根据级别汇总统计

In [146]:
frame.sum(level = 'key2')

state  Ohio     Colorado
color Green Red    Green
key2                    
1         6   8       10
2        12  14       16

In [147]:
frame.sum(level = 'color',axis = 1)

color      Green  Red
key1 key2            
a    1         2    1
     2         8    4
b    1        14    7
     2        20   10

## 21. 使用dataframe的列

In [148]:
frame = DataFrame({'a':range(7),'b':range(7,0,-1),
                  'c':['one','one','one','two','two','two','two'],
                  'd':[0,1,2,0,1,2,3]})
frame

a  b    c  d
0  0  7  one  0
1  1  6  one  1
2  2  5  one  2
3  3  4  two  0
4  4  3  two  1
5  5  2  two  2
6  6  1  two  3

In [149]:
frame2 = frame.set_index(['c','d'])
frame2

a  b
c   d      
one 0  0  7
    1  1  6
    2  2  5
two 0  3  4
    1  4  3
    2  5  2
    3  6  1

In [150]:
frame.set_index(['c','d'],drop = False)

a  b    c  d
c   d              
one 0  0  7  one  0
    1  1  6  one  1
    2  2  5  one  2
two 0  3  4  two  0
    1  4  3  two  1
    2  5  2  two  2
    3  6  1  two  3

In [151]:
frame2

a  b
c   d      
one 0  0  7
    1  1  6
    2  2  5
two 0  3  4
    1  4  3
    2  5  2
    3  6  1

In [152]:
frame2.reset_index()

c  d  a  b
0  one  0  0  7
1  one  1  1  6
2  one  2  2  5
3  two  0  3  4
4  two  1  4  3
5  two  2  5  2
6  two  3  6  1